### PR TITLE
Member resolves, IFS resolves and local `includePath` support

### PR DESCRIPTION
**Pull request by worksofliam**
### Changes

* Include and copy statements that have a member path will now resolve via the library list.
* Include and copy  statements that have a unix style path from a local file will now resolve via the `includePath` in `iproj.json`.
* Streamfiles and local sources can now support member include paths correctly (valid in RPG)

### Checklist

* [x] have tested my change
* [ ] updated relevant documentation
* [ ] Remove any/all `console.log`s I added
* [ ] eslint is not complaining
* [ ] have added myself to the contributors' list in the README
* [ ] **for feature PRs**: PR only includes one feature enhancement.

### CODE REVIEW

**Comment by chrjorgensen** [comment]
@worksofliam I was curious and went through your code for `include` and `copy` of members. And I noticed some things, that is not how the RPG compiler works (according to the [documentation](https://www.ibm.com/docs/en/i/7.3?topic=directives-copy-include)):

- If no source file is specified, you assume `QRPGLEREF` - but the compiler searches `QRPGLESRC`, not `QRPGLEREF`!
- If a member is found, you add the extension `.rpgleinc` - but the member extension is not known if shell search is performed, only if SQL is used at a later point in time.
- Extension `.rpgleinc` is only used when a member is not found and IFS files are search instead. Then IFS files with the name and extension specified is searched, or if no extension is specified, the name with extensions `.rpgle` and `.rpgleinc`.

It seems like there's some confusion about this subject, probably because of different conventions and company standards. But IMHO we should follow the rules laid out by the RPG compiler. ➡️

**Comment by worksofliam** [comment]
@chrjorgensen wow, great write up!

1. Whoops! Simple mistake on my part. I can correct to `QRPGLESRC`!
2. We use `rpgleinc` just to give the member an extension. At resolve time, the extension isn't import, but we use it just to signify it is an include.
3. I will fix up those IFS resolve inconsistencies.

Changes incoming over the next day or two with your suggestions. Thanks you!!

**Comment by worksofliam** [comment]
@chrjorgensen Based on your feedback, the only change I made is to use `QRPGLESRC` instead of `QRPGLEREF`.

The plan is to first get member resolves working, and then move onto IFS/local.

**Comment by worksofliam** [comment]
@chrjorgensen Further updates to improve the code and allow cross file system resolves and cleaner code.

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
@worksofliam 

I tested by creating a library TESTPR225 and using the following RPGLE main source with `/copy` and `/include` statements from the documentation:

```rpgle
**free

/COPY MBR1
/INCLUDE SRCFIL,MBR2
/COPY TESTPR225/SRCFIL,MBR3
/INCLUDE TESTPR225/"SRC>3","MBR¬3"
/COPY /dir1/dir2/file.rpg
/COPY /dir1/dir2/file
/COPY dir1/dir2/file.rpg¬3"
/COPY "ifs file containing blanks"
/COPY 'ifs file containing blanks'
```

I tested the member / streamfile resolve by putting the cursor on the name and press `Alt-F12` to peek the content of the include file.

I noted the following:

1. The copy of `MBR1` worked fine, member found was `TESTPR225/QRPGLESRC/MBR1.RPGLE`.
2. The include of `SRCFIL/MBR2` worked fine, member found was `TESTPR225/SRCFIL/MBR2.RPGLE`.
3. The copy of `TESTPR225/SRCFIL/MBR3` worked fine, member found was `TESTPR225/SRCFIL/MBR3.RPGLE`.
4. The include of `TESTPR225/"SRC>3","MBR¬3"` could not be resolved.  Compile found it.
5. The IFS copy of `/dir1/dir2/file.rpg` could not be resolved. Compile found it.
6. The IFS copy of `/dir1/dir2/file` could not be resolved.  Compile found it.
7. The IFS copy of `dir1/dir2/file.rpg` could not be resolved.  Compile found it.
8. The IFS copy of `"ifs file containing blanks"` could not be resolved.  Compile found it.
9. The IFS copy of `'ifs file containing blanks'` could not be resolved.  Compile found it.
10. The member resolve worked both when `TESTPR225` was the current library or in the library list (I had it as the last library in the list with a prior library also containing a source file called `SRCFIL`).

Changing the current library also didn't affect the resolve cache - I had to disconnect and reconnect to make the resolve work, after first having another current library than `TESTPR225`.

**Should there be some command or other way of resetting the cache? Would be nice for the user to be able to do that, when the cache has old or obsolete content... or at least it should be cleared when the current library or library list changes...**

**Comment by worksofliam** [comment]
@chrjorgensen 

- [x] The include of TESTPR225/"SRC>3","MBR¬3" could not be resolved. Compile found it.
    - No plans to implement this style.
- [x] The IFS copy of /dir1/dir2/file.rpg could not be resolved. Compile found it.
    - This has been done. Explicit path started from root (`/`) will now resolve.
- [ ] The IFS copy of /dir1/dir2/file could not be resolved. Compile found it.
    - To be finished. Not sure how this will work yet.
- [x] The IFS copy of dir1/dir2/file.rpg could not be resolved. Compile found it.
    - This has been done. If not root (`/`) is provided, it will search relative of the working directory (which can be changed in the IFS path)
- [ ] The IFS copy of "ifs file containing blanks" could not be resolved. Compile found it.
    - Support for double quotes not yet finished.
- [x]The IFS copy of 'ifs file containing blanks' could not be resolved. Compile found it.
    - I think this should work now.

> Changing the current library also didn't affect the resolve cache - I had to disconnect and reconnect to make the resolve work, after first having another current library than TESTPR225.

You can simple close and re-open the member/IFS/local file now.

I will post again when the remaining items are complete.

**Comment by worksofliam** [comment]
Great progress has been made.

image

@chrjorgensen Back to you for another review whenever you can.

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
I did some further testing on the same test source:

:x: The copy of `TESTPR225/SRCFIL,MBR3` no longer worked - your code probably thinks it's a UNIX path, when it's actually a QSYS path.
✔️  The IFS copy of `/dir1/dir2/file.rpg` worked now.
:x: The IFS copy of `/dir1/dir2/file` worked - partially. Could only be resolved, when exact filename existed, but no seatch for suffixes `.rpgle` and `.rpgleinc` is performed.
✔️ The IFS copy of `dir1/dir2/file.rpg` worked now.
❌ I could not get the IFS copy of `"ifs file containing blanks"` to work.
❌ I could not get the IFS copy of `'ifs file containing blanks'` to work.

You should probably also take into consideration the rules for when the compiler can't determine the filesystem specified, as described here: [https://www.ibm.com/docs/en/i/7.3?topic=files-include](https://www.ibm.com/docs/en/i/7.3?topic=files-include). Then both QSYS and IFS will be searched as documented. I see that you already commented on this in the code, but it seems not implemented yet?

**Comment by chrjorgensen** [comment]
@worksofliam IMHO the RPG /COPY documentation is confusingly organized - the [INCLUDE FILES page](https://www.ibm.com/docs/en/i/7.3?topic=files-include) seems the most important. It was probably added when RPG source code was allowed in the IFS.

An example that I was not really aware of:
`/COPY mylib/myfile,mymbr` can mean 
1. in the QSYS filesystem library `mylib` with source file `myfile` and member `mymbr` **OR**
2. in the IFS (root) filesystem subdirectory `mylib` in the current working directory with streamfile `myfile,mymbr`

The page above has more examples.

So both QSYS and IFS filesystem are searched by the compiler unless the filename is in quotes.

I won't even go into the IFS include path using `INCDIR` parameter or `RPGINCDIR` environment variable...!

**Comment by chrjorgensen** [comment]
One test still missing is `/copy` files in `includePath` - coming up later...

File: extension/server/src/server.ts
Comment (worksofliam) [reply extension/server/src/server.ts:199]: We need to wait for 2.0.0 of vscode-ibmi to search for multiple extensions.

File: extension/server/src/server.ts
Comment (chrjorgensen) [inline extension/server/src/server.ts:138]: I'm not too happy about the new test for `,`:

```javascript
const isUnixPath = hasQuotes || (includeString.includes(`/`) && !includeString.includes(`,`));
```

Considering a `,` in a filename a QSYS member does not quite follow the compiler behaviour - and someone may be burnt by this. Ideally we should follow [the rules](https://www.ibm.com/docs/en/i/7.3?topic=files-include) as closely as possible and search in both filesystems, unless the name can only exist in one filesystem.

I don't think we should worry about performance too much. Most times the include files will be in the expected place and will be found quickly - and if they are in another filesystem, a little wait is okay. Better right than quick! 😆

Comment (worksofliam) [reply extension/server/src/server.ts:138]: > Considering a , in a filename a QSYS member does not quite follow the compiler behaviour - and someone may be burnt by this.

You are likely right, but I am happy to say that vscode-rpgle won't support every method of including sources from the get go. For example, there is no plan for me to add support for this as of right now: `/COPY "A/b",mymbr`

I also really don't want to add code to search members first and then fall back to IFS files. This is a personal preference because I am worried about performance (sorry!) and I remember when I tried to write this a few years ago it didn't go so well.

I would like our code to determine if this is an IFS include OR a member include and then try to bring them in. I will try and think of another method instead of checking for `,` 😆

**Comment by worksofliam** [comment]
@chrjorgensen I know you're about to depart for holiday, but whenever you're back...


✅ The copy of TESTPR225/SRCFIL,MBR3 no longer worked
This is fixed.
✔️ The IFS copy of /dir1/dir2/file.rpg worked now.
✅ The IFS copy of /dir1/dir2/file worked - partially. Could only be resolved, when exact filename existed, but no seatch for suffixes .rpgle and .rpgleinc is performed.
Left the current logic in until vscode-ibmi 2.0.0, which will allow us to search for many names.
✔️ The IFS copy of dir1/dir2/file.rpg worked now.
✅ I could not get the IFS copy of "ifs file containing blanks" to work.
This is fixed. Share what your include directive looks like if it fails again.
✅ I could not get the IFS copy of 'ifs file containing blanks' to work.
This is fixed. Share what your include directive looks like if it fails again.

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
@worksofliam I made a new review of the source and include statements:

1. :question: The copy of `MBR1` worked almost, member found was `TESTPR225/QRPGLESRC/MBR1.rpgleinc` - but member extension actually was `TXT`.
2. :question: The include of `SRCFIL/MBR2` worked almost, member found was `TESTPR225/SRCFIL/MBR2.rpgleinc` - but member extension actually was `RPGLE`.
3. :question: The copy of `TESTPR225/SRCFIL/MBR3` worked almost, member found was `TESTPR225/SRCFIL/MBR3.rpgleinc` - but member extension actually was `RPGLE`.
4. :heavy_check_mark: The IFS copy of `/dir1/dir2/file.rpg` worked fine.
5. :question: The IFS copy of `/dir1/dir2/file` only worked, when the filename was `file` - but not with `file.rpgle` or `file.rpgleinc`. Can we include multi file search now?
6. :heavy_check_mark: The IFS copy of `dir1/dir2/file.rpg` worked fine.
7. :x: The IFS copy of `"ifs file containing blanks"` could not be resolved, even though a file named `ifs file containing blanks` was put in the current working directory.
8. :x: The IFS copy of `'ifs file containing blanks'` could not be resolved, even though a file named `ifs file containing blanks` was put in the current working directory.

I think we're very close to having it - only minor issues except for the question of multi file search... 👍 

edit from barry: added numbers to make it easier to reference

**Comment by worksofliam** [comment]
Thanks for testing @chrjorgensen!

For 1, 2 & 3: Right now, when including a member we default to `rpgleinc` as the extension. Our APIs don't provide the extension (always simply returns `.MBR`) so I decided to use `.rpgleinc` for now. Adding an additional call to get the valid extension is expensive and didn't seem worthy for now. While this may confuse our developers, I think initially this is something we can start with.

For 5: I tested this and it seemed to work as you described:

image

For 7 & 8: I mentioned previously that this style of include (paths with spaces) won't be supported in this PR (a risk I am also willing to take). It looks like there is a bug in the `memberResolve` API that will need fixing before this can work. I don't want to hold up this PR.

**Comment by chrjorgensen** [comment]
@worksofliam IMHO it would be better to have no extension for members - since the extension is not specified in the `/COPY` statement and the extension actually is not part of the member name.

Instead of showing 
billede
it should just show `/TESTPR225/QRPGLESRC/MBR1` - is this difficult to implement?

I have reviewed your [PR #1482](https://github.com/halcyon-tech/vscode-ibmi/pull/1482) and made a suggestion...

**Review by chrjorgensen — APPROVED** [review]
@worksofliam I just did a final test: The hover is great now, 5 and 7-8 are still not working.

But we can fix this later, let's get this in now! Okay?

**Comment by worksofliam** [comment]
@chrjorgensen 5 isn't working for you? That is odd! I actually wonder if we should take a look at that together before we merge this.

**Comment by worksofliam** [comment]
@chrjorgensen I went ahead and fixed the issue where the path name was being cut off as you discovered! Whenever you have the time, take another look with my latest PR!

**Review by chrjorgensen — APPROVED** [review]
@worksofliam Everything is working as expected now - even test 5, which was due to a mistake on my part!

Looking good! Approved! Merge at any time! 😄

**Pull request merged**
